<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8"/>
    <meta name="viewport" content="width=device-width, user-scalable=no, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1"/>
    <title>For Jupyter Users &#8211; Gene Pattern Notebook</title>
    <link rel='stylesheet' id='font-awesome-css' href='https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css?ver=4.7.4' type='text/css' media='all'/>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
    <script type='text/javascript' src='../wp-includes/js/jquery/jquery.js?ver=1.12.4'></script>
    <script type='text/javascript' src='../wp-includes/js/jquery/jquery-migrate.min.js?ver=1.4.1'></script>
    <script type='text/javascript' src='../wp-content/themes/genepattern/js/onepagenav.js?ver=4.7.4'></script>
    <script type='text/javascript' src='../wp-content/themes/genepattern/js/stickykit.js?ver=4.7.4'></script>
    <script type='text/javascript' src='../wp-content/themes/genepattern/js/scripts.js?ver=4.7.4'></script>
    <script type='text/javascript'>
    /* When the user clicks on the button, 
    toggle between hiding and showing the dropdown content */
    function myFunction() {
        document.getElementById("myDropdown").classList.toggle("show");
    }
    /* When the user clicks on the button, 
    toggle between hiding and showing the dropdown content */
    function myFunction2() {
        document.getElementById("myDropdown2").classList.toggle("show");
    }
    // Close the dropdown if the user clicks outside of it
    window.onclick = function(e) {
      if (!e.target.matches('.dropbtn')) {
        var myDropdown = document.getElementById("myDropdown2");
          if (myDropdown.classList.contains('show')) {
            myDropdown.classList.remove('show');
          }
        var myDropdown = document.getElementById("myDropdown");
          if (myDropdown.classList.contains('show')) {
            myDropdown.classList.remove('show');
          }
      }
    }
    </script>
    <!-- Global Site Tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-106705148-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments)};
      gtag('js', new Date());

      gtag('config', 'UA-106705148-1');
    </script>
    <link rel="stylesheet" href="../wp-content/themes/genepattern/dropdown.css" type="text/css" media="screen" />
    <link rel="stylesheet" href="../wp-content/themes/genepattern/style.css" type="text/css" media="screen"/>
    <style type="text/css">
        table td {
            font-size: 0.8em;
            text-align: left;
        }
    </style>
</head>

<body class="page-template-default page page-id-9">
    <div id="wrapper">
        <header id="header">
            <div class="container cf">
                <a href=".." id="logo"><img src="../wp-content/themes/genepattern/images/logo.png" alt="Gene Pattern Notebook" /></a>
               <nav>
                    <a href="#" id="mobile-menu"><i class="fa fa-bars"></i></a>
                    <ul id="menu-main-menu" class="menu">
                        <li id="dropdown" class="menu-item">
                            <ol class="dropdown">
                                <li class="dropbtn" onclick="myFunction()" style="margin-left: 0%">Guides</li>
                                <hr style="height: 18px; visibility: hidden;" id="mobile">
                                <div class="dropdown-content" id="myDropdown">
                                    <a href="../user-guide/">Getting Started</a>
                                    <hr style="margin: 1%">
                                    <a href="../basic-features/">Basic Features</a>
                                    <a href="../public-notebooks/">Public Notebooks</a>
                                    <a href="../publishing-sharing/">Publishing & Sharing Notebooks</a>
                                    <a href="../new-features/">New Features</a>
                                    <hr style="margin: 1%">
                                    <a href="../programmatic/">Programmatic Features</a>
                                    <a href="../install/">Local Installation</a>
                                    <a href="../for-jupyter-users/">Jupyter Users</a>
                                </div>         
                            </ol>
                        </li>
                        <li id="menu-item-18" class="menu-item"><a href="../tutorials/">Videos/Tutorials</a></li>
                        <li id="menu-item-17" class="menu-item menu-item- ntype-post_type menu-item-object-page menu-item-17"><a href="../faq/">FAQ</a></li>
                        <li id="dropdown" class="menu-item">
                            <ol class="dropdown">
                                <li class="dropbtn" onclick="myFunction2()" style="margin-left: 0%">Resources</li>
                                <hr style="height: 18px; visibility: hidden;" id="mobile">
                                <div class="dropdown-content" id="myDropdown2">
                                    <a href="../community/">Community</a>
                                    <a href="../releases/">Releases</a>
                                    <a href="../example-notebooks/">Example Notebooks</a>
                                    <a href="../cheat-sheet/">Cheat Sheet</a>
                                </div>         
                            </ol>
                        </li>
                        <!-- <li id="menu-item-16" class="menu-item menu-item-type-post_type menu-item-object-page menu-item-16"><a href="../documentation/">Documentation</a></li> -->
                        <li id="menu-item-16" class="menu-item menu-item-type-post_type menu-item-object-page menu-item-16"><a href="../about/">About</a></li>
                        <li id="menu-item-16" class="menu-item menu-item-type-post_type menu-item-object-page menu-item-16"><a href="../contact/">Contact</a></li>
                        <li id="menu-item-51" class="button" style="font-weight: 300"><a href="https://notebook.genepattern.org/hub/login">Login/Register</a></li>
                    </ul>			
                </nav>
            </div>
        </header>
        <div id="main" class="container cf">
            <section>
                <h1>GenePattern Notebook for Jupyter Users </h1>
                <hr>
                <h3>1. GenePattern Notebook Extension</h3>
                <p>GenePattern Notebook is an extension for Jupyter. It provides a variety of tools for interacting with the powerful GenePattern platform from within a Jupyter Notebook environment. These tools include:</p>
                <ul>
                    <li>Visual widgets for running bioinformatic analyses and submitting these analyses to a GenePattern server.</li>
                    <li>Tools for visually sending results to downstream steps in a bioinformatic workflow.</li>
                    <li>A WYSIWYG rich text editor for Markdown cells.</li>
                    <li>Jupyter magics for creating GenePattern widgets.</li>
                    <li>A Python library for programmatically working with the results of the analyses.</li>
                </ul>
                <p>To use the GenePattern Notebook extension, you will need to <a href="https://notebook.genepattern.org">register
                    for an account</a>. GenePattern usage is free and the software is open source under a BSD-style license.</p>

                <h3>2. What is GenePattern?</h3>
                <p>GenePattern is a powerful platform for reproducible bioinformatic research. It provides hundreds of analytical tools for the analysis of gene expression (<a href="https://genepattern.org/rna-seq-analysis">RNA-seq</a> and <a
                href="https://genepattern.org/gene-expression-analysis">microarray</a>), <a href="https://genepattern.org/variant-and-copy-number-analysis">sequence variation and copy number</a>, <a href="https://genepattern.org/proteomics">proteomic</a>, <a
                href="https://genepattern.org/flow-cytometry">flow cytometry</a>, and network analysis.</p>
                <p>For more information <a href="https://genepattern.org">visit our main GenePattern page</a>. For introductory information about GenePattern, please see our <a href="https://software.broadinstitute.org/cancer/software/genepattern/quick-start">10-minute tutorial</a> and/or check out one of our <a href="https://software.broadinstitute.org/cancer/software/genepattern/video-tutorials">video tutorials</a>.</p>

                <h3>3. GenePattern Cells</h3>
                <p>GenePattern cells are Python code cells which are automatically executed, and which display a visual widget for submitting GenePattern analyses
                or working with the results of these analyses. These cells are displayed as a new cell type, and can be added either through the Jupyter menu or
                through the GenePattern Notebook search functionality.</p>
                <p><img alt="" src="../wp-content/uploads/2017/02/content_screen_shot_2015-10-22_at_11_50_01.png" style="width: 600px; height: 251px; border-width: 1px; border-style: solid;"/></p>
                <p><img alt="" src="../wp-content/uploads/2017/02/content_screen_shot_2015-08-24_at_10_30_39.png" style="width: 600px; height: 296px; border-width: 1px; border-style: solid;"/></p>
                <p>GenePattern cells come in three types: authentication, analysis cells and UI Builder cells.</p>
                <p><strong>A. Authentication cells</strong> are used to authenticate the notebook user with a GenePattern server. One of these cells must be added to a notebook before Analysis cells becomes available. If a GenePattern cell is added to a notebook before authentication has happened, it will automatically become one of these cells.</p>
                <p><img alt="" src="../wp-content/uploads/2017/02/content_screen_shot_2015-08-24_at_10_27_12.png" style="width: 600px; height: 221px; border-width: 1px; border-style: solid;"/></p>
                <p><strong>B. Analysis cells</strong> each represent a particular type of analysis on the GenePattern platform. Hundreds of diferent analyses are available. These cells will present a form to the user, accepting data or other parameters. Once one of these forms has been filled out, the user can then run the analysis by clicking the Run button. This will submit the analysis to the GenePattern server and display the results.</p>
                <p><img alt="" src="../wp-content/uploads/2017/02/content_screen_shot_2015-08-24_at_10_32_11.png" style="width: 600px; height: 324px; border-width: 1px; border-style: solid;"/></p>
                <p><strong>C. UI Builder cells</strong> are a way to display any Python function or method call as an interactive widget. This will render the parameters of the function as a web form. The UI Builder will use any existing docstring for the function as a description, will infer parameter types from default values and will display parameter annotations as helpful text near each input.</p>
                <p><img alt="" src="../programmatic/ui_builder_1.jpg" style="width: 100%;"></p>
                <p>Since GenePattern cells are also Python code cells, all changes to the information in these widgets is reflected in the code and will be saved in
                the notebook&#39;s ipynb file. Similarly, sending one of these notebook files to a user without the GenePattern Notebook extension installed, will result in a complete notebook, albeit one where the cells with visual widgets are instead displayed as code.</p>
                <p>To view the code backing a GenePattern cell, simply click the &quot;Gear&quot;&nbsp;button found on a GenePattern widget in the upper right
                corner and select &quot;Toggle Code.&quot;</p>

                <h3>4. UI Builder</h3>
                <p>The UI Builder is a way to display any Python function or method call as an interactive widget. This will render the parameters of the
                    function as a web form which can be filled out with numbers, string literals or with references to existing Python variables. Clicking the
                    <em>Run</em> button on the widget will validate the form and execute the function using the supplied values.</p>
                <p>The UI Builder will use any existing docstring for the function as a description in the widget, will infer parameter types from default
                    values and will display parameter annotations as helpful text near each input.</p>
                <p>To use render a function, simply import the <em>UIBuilder</em> class from the <em>nbtools</em> package and pass the function to its
                    constructor.</p>
                <p><a href="/programmatic/#ui-builder">For more information see the Programmatic Features page</a></p>
                <p><img alt="" src="call-widget.jpg" style="width: 600px; border: 1px solid;"/></p>

                <h3>5. GenePattern Python Library</h3>
                <p>Programmatic access to all GenePattern jobs and results is also automatically available using GenePattern&#39;s Python library. Any job executed in a notebook can afterward be referenced using <code><em>job</em></code> followed by the job number. For example, if a job is number 1170434,
                the variable name would be <code><em>job1170434</em></code>. Entering this into a code cell will return a reference to a GPJob object.</p>
                <ul>
                    <li>For more information on using the GenePattern Python library, see the <a href="https://genepattern.org/programmers-guide#_Using_GenePattern_from_Python">GenePattern Programmer&#39;s
                    Guide</a>.
                    </li>
                    <li>A notebook providing a <a href="https://github.com/genepattern/example-notebooks/blob/master/GenePattern%20Python%20Tutorial.ipynb">tutorial
                    for using the GenePattern Python Library</a> is also available.
                    </li>
                </ul>
                <p>Code examples of how to reference GenePattern jobs or GenePattern result files are available in the notebook by clicking a job result and selecting &quot;Send to Code&quot; in the menu.</p>
                <p><img alt="" src="../wp-content/uploads/2017/02/content_screen-shot-2015-10-15-at-13_50.jpg" style="width: 600px; height: 159px; border-width: 1px; border-style: solid;"/></p>
                <p>GPJob objects can also be programmatically created and executed without the need for a GenePattern cell or widget. For information on how to do
                this see the <a href="https://genepattern.org/programmers-guide#_Using_GenePattern_from_Python">GenePattern Programmer&#39;s Guide</a>&nbsp;or the <a href="https://github.com/genepattern/example-notebooks/blob/master/GenePattern%20Python%20Tutorial.ipynb">GenePattern Python Tutorial</a> notebook.</p>

                <h1>Python Packages</h1>
                <p>The GenePattern Notebook Repository comes with a wide selection of Python packages installed in the base environment. Users can also supplement this list by installing their own packages. The base environment includes:</p>
                <table class="table table-striped">
                    <tr>
                        <td><strong>anndata</strong></td>
                        <td>argcomplete</td>
                        <td>asnrypto</td>
                        <td>attrs</td>
                        <td>Babel</td>
                    </tr>
                    <tr>
                        <td>backcall</td>
                        <td>beautifulsoup</td>
                        <td><strong>biopython</strong></td>
                        <td>bleach</td>
                        <td><strong>bokeh</strong></td>
                    </tr>
                    <tr>
                        <td>boto</td>
                        <td>botocore</td>
                        <td>cachetools</td>
                        <td><strong>ccalnoir</strong></td>
                        <td>certifi</td>
                    </tr>
                    <tr>
                        <td>cffi</td>
                        <td>chardet</td>
                        <td>cloudpickle</td>
                        <td>CommonMark</td>
                        <td>crc32c</td>
                    </tr>
                    <tr>
                        <td>crcmod</td>
                        <td>cryptography</td>
                        <td><strong>cuzcatlan</strong></td>
                        <td>cycler</td>
                        <td><strong><a href="https://github.com/cytoscape/cytoscape-jupyter-widget" target="_blank">cyjupyter</a></strong></td>
                    </tr>
                    <tr>
                        <td>Cython</td>
                        <td>cytoolz</td>
                        <td>dask</td>
                        <td>dcplib</td>
                        <td>decorator</td>
                    </tr>
                    <tr>
                        <td>defusedxml</td>
                        <td>dill</td>
                        <td>docutils</td>
                        <td>entrypoints</td>
                        <td>enum34</td>
                    </tr>
                    <tr>
                        <td>fastcache</td>
                        <td>future</td>
                        <td><strong>genefab</strong></td>
                        <td><strong>genepattern-notebook</strong></td>
                        <td><strong>genepattern-python</strong></td>
                    </tr>
                    <tr>
                        <td>gmpy2</td>
                        <td>google-auth</td>
                        <td>google-auth-oauthlib</td>
                        <td>h5py</td>
                        <td><strong>hca</strong></td>
                    </tr>
                    <tr>
                        <td><strong>humanfriendly</strong></td>
                        <td>idna</td>
                        <td><strong><a href="https://github.com/igvteam/igv-jupyter#igvjs-jupyter-extension" target="_blank">igv-jupyter</a></strong></td>
                        <td>ijson</td>
                        <td>imageio</td>
                    </tr>
                    <tr>
                        <td>ipykernel</td>
                        <td>ipython</td>
                        <td>ipython-genutils</td>
                        <td><strong>ipywidgets</strong></td>
                        <td>jedi</td>
                    </tr>
                    <tr>
                        <td>Jinja2</td>
                        <td>jmespath4</td>
                        <td>joblib</td>
                        <td>jsonpointer</td>
                        <td>jsonschema</td>
                    </tr>
                    <tr>
                        <td>jupyter</td>
                        <td>jupyter-client</td>
                        <td>jupyter-console</td>
                        <td>jupyter-core</td>
                        <td><strong>jupyter-wysiwyg</strong></td>
                    </tr>
                    <tr>
                        <td>kiwisolver</td>
                        <td>llvmlite</td>
                        <td>louvain</td>
                        <td>MarkupSafe</td>
                        <td><strong>matplotlib</strong></td>
                    </tr>
                    <tr>
                        <td>memory-profiler</td>
                        <td>mistune</td>
                        <td>mock</td>
                        <td>mpmath</td>
                        <td>natsort</td>
                    </tr>
                    <tr>
                        <td>nbconvert</td>
                        <td>nbformat</td>
                        <td><strong>nbtools</strong></td>
                        <td><strong>ndex2</strong></td>
                        <td><strong>networkx</strong></td>
                    </tr>
                    <tr>
                        <td>notebook</td>
                        <td><strong>numba</strong></td>
                        <td>numexpr</td>
                        <td><strong>numpy</strong></td>
                        <td>oauthlib1</td>
                    </tr>
                    <tr>
                        <td>olefile</td>
                        <td>opencv-python</td>
                        <td>packaging</td>
                        <td><strong>pandas</strong></td>
                        <td>pandocfilters</td>
                    </tr>
                    <tr>
                        <td>parso</td>
                        <td>patsy</td>
                        <td>pexpect</td>
                        <td>pickleshare</td>
                        <td>Pillow</td>
                    </tr>
                    <tr>
                        <td><strong>plotly</strong></td>
                        <td>prometheus-client</td>
                        <td>prompt-toolkit</td>
                        <td>protobuf</td>
                        <td>psutil2</td>
                    </tr>
                    <tr>
                        <td>ptyprocess</td>
                        <td>puremagic</td>
                        <td><strong>py2cytoscape</strong></td>
                        <td>pyasn5</td>
                        <td>pyasnmodules</td>
                    </tr>
                    <tr>
                        <td>pycairo</td>
                        <td>pycparser</td>
                        <td>pydot1</td>
                        <td>pydotplus</td>
                        <td>Pygments</td>
                    </tr>
                    <tr>
                        <td>PyJWT1</td>
                        <td>pyparsing</td>
                        <td>pyrsistent</td>
                        <td><strong>pysam</strong></td>
                        <td>pysolr</td>
                    </tr>
                    <tr>
                        <td>python-dateutil</td>
                        <td>python-igraphpost</td>
                        <td>pytz</td>
                        <td>PyWavelets</td>
                        <td>PyYAML</td>
                    </tr>
                    <tr>
                        <td>pyzmq</td>
                        <td>qtconsole</td>
                        <td>requests</td>
                        <td>requests-oauthlib</td>
                        <td>requests-toolbelt</td>
                    </tr>
                    <tr>
                        <td><strong>rpy2</strong></td>
                        <td>rsa</td>
                        <td>s3transfer</td>
                        <td><strong>scanpy</strong></td>
                        <td>scikit-image</td>
                    </tr>
                    <tr>
                        <td><strong>scikit-learn</strong></td>
                        <td><strong>scipy</strong></td>
                        <td><strong>seaborn</strong></td>
                        <td>Send2Trash</td>
                        <td>simplegeneric</td>
                    </tr>
                    <tr>
                        <td>six</td>
                        <td>soupsieve</td>
                        <td>SQLAlchemy</td>
                        <td>statsmodels</td>
                        <td>sympy</td>
                    </tr>
                    <tr>
                        <td>tables</td>
                        <td>tenacity</td>
                        <td>terminado</td>
                        <td>testpath</td>
                        <td>toolz</td>
                    </tr>
                    <tr>
                        <td>tornado</td>
                        <td>tqdm</td>
                        <td>traitlets</td>
                        <td>tweak</td>
                        <td>tzlocal</td>
                    </tr>
                    <tr>
                        <td>umap-learn</td>
                        <td>urllib</td>
                        <td>validators</td>
                        <td>vincent</td>
                        <td>wcwidth</td>
                    </tr>
                    <tr>
                        <td>webencodings</td>
                        <td>widgetsnbextension</td>
                        <td>xlrd</td>
                    </tr>
                </table>



            </section>
            <aside>
            </aside>
        </div>
        <footer id="footer">
            <div class="container">
                <p>GenePattern Notebook usage is free and the software is open source
                    under a BSD-style license.</p>
                <p>© 2019 Regents of the University of California and the Broad Institute. </p>
            </div>
        </footer>
    </div>
    <script type='text/javascript' src='../wp-includes/js/wp-embed.min.js?ver=4.7.4'></script>

</body>
</html>
